### PR TITLE
fix: Migration order due to cherry which went astray

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

This PR fixes a Alembic migration conundrum when upgrading from `3.0` to `3.1` due to a [picked cherry](https://github.com/apache/superset/commit/88e6f22180b61f1c3ba15134f2f64064a448ff92) which somehow went astray. 

The Alembic visions chain on the various branches is:

#### master

- `ec54aca4c8a2`
- `317970b4400c`
- `4b85906e5b91`

#### 3.0

- `ec54aca4c8a2`
- `4b85906e5b91`

#### 3.1 

- `ec54aca4c8a2`
- `317970b4400c`
- `4b85906e5b91`

i.e., per the previously mentioned cherry in 3.0 revision `4b85906e5b91` somehow became rewired from `317970b4400c` to `ec54aca4c8a2` meaning the `317970b4400c` was never present in 3.0. 

This PR:

1. Adds safeguards to `317970b4400c`.
2.  Adds revision `b7851ee5522f`—which should also be cherry-picked into `3.1`—which merely replays `317970b4400c` for both `master` and `3.1`.

These two steps ensure that if your upgrading/downgrading from either `master` or `3.0` that `317970b4400c` will run if and only if the migration has not been run previous—determined by the presence/absence of the column respectively.

### TESTING INSTRUCTIONS

CI.

### ADDITIONAL INFORMATION

- [x] Has associated issue: Fixes https://github.com/apache/superset/issues/26163
- [ ] Required feature flags:
- [ ] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [x] Migration is atomic, supports rollback & is backwards-compatible
  - [x] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by michael-s-molina — APPROVED** [review]
LGTM

File: superset/migrations/shared/utils.py [hunk 1]
```diff
@@ -43,11 +43,9 @@ def table_has_column(table: str, column: str) -> bool:
     :param column: A column name
     :returns: True iff the column exists in the table
     """
-    config = op.get_context().config
-    engine = engine_from_config(
-        config.get_section(config.config_ini_section), prefix="sqlalchemy."
-    )
-    insp = reflection.Inspector.from_engine(engine)
+
+    insp = inspect(op.get_context().bind)
```
Comment (john-bodley) [inline superset/migrations/shared/utils.py:47]: The `table_has_column` method was returning `False` for PostgreSQL within CI even though the columns existed. The proposed update is from the same [Stack Overflow](https://stackoverflow.com/questions/24082542/check-if-a-table-column-exists-in-the-database-using-sqlalchemy-and-alembic) post, however it's from 2022 (as opposed to 2018) and has more votes.

**Comment by eschutho** [comment]
Thanks @john-bodley, this looks good to me too. I outlined what the "after" would look like as I was reading through this, so pasting it here for anyone else interested. (pls correct me if I missed anything)

master
- ec54aca4c8a2
- 317970b4400c
- 4b85906e5b91
- b7851ee5522f (replays 317970b4400c)

3.0
- ec54aca4c8a2
- 4b85906e5b91

3.1
- ec54aca4c8a2
- 317970b4400c
- 4b85906e5b91

3.1.x
- ec54aca4c8a2
- 317970b4400c
- 4b85906e5b91
- b7851ee5522f (replays 317970b4400c)

---------------------------------------------------------------------------
So someone upgrading from 3.0 to 3.1.x for example would get:
- ec54aca4c8a2
- 4b85906e5b91
- b7851ee5522f (replays 317970b4400c)

**Review by eschutho — APPROVED** [review]

**Comment by john-bodley** [comment]
Thanks @eschutho for adding this—I was wondering whether I should have added something similar. I did update your comment to include `b7851ee5522f` in `3.1`. We likely could just remove the `3.1.x` section as well.

**Review by justinpark — APPROVED** [review]

**Comment by eschutho** [comment]
Great, I updated the comment for anyone else reading this, removed 3.1 ( I was thinking 3.1.0 specifically, but this works), and I showed where the `317970b4400c` wouldn't run. lmk if that comment makes more sense, or feel free to edit.

**Pull request merged**